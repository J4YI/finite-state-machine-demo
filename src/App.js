import React from 'react'

import Error from './components/Error'
import Idle from './components/Idle'
import Loading from './components/Loading'
import User from './components/User'
import { Main, Text } from './styled'
import { fetchUser } from './repository'
import Machine, {
  events,
  actions,
  states,
  withMachine
} from './machine'

export const machine = {
  initial: states.IDLE,
  states: {
    [states.IDLE]: {
      on: {
        [events.FETCH]: states.FETCHING
      }
    },
    [states.FETCHING]: {
      on: {
        [events.FULFILL]: states.SUCCESS,
        [events.FAIL]: states.FAILURE
      },
      action: actions.FETCH_USER
    },
    [states.SUCCESS]: {
      on: {
        [events.FETCH]: states.FETCHING
      }
    },
    [states.FAILURE]: { on: {} }
  }
}

export class App extends React.Component {
  constructor (props) {
    super(props)
    this.state = { user: {} }

    // Add commands listeners
    props.commands({
      [actions.FETCH_USER]: () => this.fetchUser()
    })
  }

  async fetchUser () {
    try {
      const user = await fetchUser()

      this.setState({ user }, () => {
        this.props.transition(events.FULFILL)
      })
    } catch (e) {
      this.props.transition(events.FAIL)
    }
  }

  renderFromMachine () {
    switch (this.props.machineState) {
      case states.IDLE: return <Idle
        transition={this.props.transition}
      />
      case states.FETCHING: return <Loading />
      case states.SUCCESS: return <User
        user={this.state.user}
        transition={this.props.transition}
      />
      case states.FAILURE:
      default:
        return <Error />
    }
  }

  render () {
    return (
      <Main>
        <Text>State: {this.props.machineState}</Text>
        {this.renderFromMachine()}
      </Main>
    )
  }
}

export default withMachine(Machine(machine))(App)
