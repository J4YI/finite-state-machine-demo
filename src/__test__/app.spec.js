/* global it, expect */
import React from 'react'
import renderer from 'react-test-renderer'

import { machine, App } from '../App'
import Machine, { withMachine } from '../machine'

it('should render all the possible states', () => {
  Object.keys(machine.states).forEach((initial) => {
    const stateMachine = { ...machine, initial }
    const Automata = withMachine(Machine(stateMachine))(App)
    const wrapper = renderer.create(<Automata />)

    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
