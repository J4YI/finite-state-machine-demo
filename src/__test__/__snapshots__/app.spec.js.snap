// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render all the possible states 1`] = `
<main
  className="sc-bwzfXH ceTrXN"
>
  <p
    className="sc-htpNat cdeVmq"
  >
    State: 
    IDLE
  </p>
  <p
    className="sc-htpNat cdeVmq"
  >
    Let's do a search 
    <span
      aria-label="emoji"
      role="img"
    >
      🕵🏻‍♂️
    </span>
  </p>
  <button
    className="sc-bdVaJa iSLsWi"
    onClick={[Function]}
  >
    Find new one!
  </button>
</main>
`;

exports[`should render all the possible states 2`] = `
<main
  className="sc-bwzfXH ceTrXN"
>
  <p
    className="sc-htpNat cdeVmq"
  >
    State: 
    FETCHING
  </p>
  <p
    className="sc-htpNat cdeVmq"
  >
    <span
      aria-label="emoji"
      role="img"
    >
      👀
    </span>
  </p>
  <button
    className="sc-bdVaJa hwcSD"
    disabled={true}
  >
    Searching...
  </button>
</main>
`;

exports[`should render all the possible states 3`] = `
<main
  className="sc-bwzfXH ceTrXN"
>
  <p
    className="sc-htpNat cdeVmq"
  >
    State: 
    SUCCESS
  </p>
  <p
    className="sc-htpNat cdeVmq"
  >
    <span
      aria-label="emoji"
      role="img"
    />
     
  </p>
  <button
    className="sc-bdVaJa iSLsWi"
    onClick={[Function]}
  >
    Find another
  </button>
</main>
`;

exports[`should render all the possible states 4`] = `
<main
  className="sc-bwzfXH ceTrXN"
>
  <p
    className="sc-htpNat cdeVmq"
  >
    State: 
    FAILURE
  </p>
  <p
    className="sc-htpNat cdeVmq"
  >
    Something went wrong 
    <span
      aria-label="emoji"
      role="img"
    >
      💩
    </span>
  </p>
</main>
`;
